From d78e2a4e7de84e9d5ca978201aa0590f0ab9f8a6 Mon Sep 17 00:00:00 2001
From: Dan Streetman <ddstreet@canonical.com>
Date: Mon, 27 Jul 2020 12:34:38 -0400
Subject: [PATCH 4/5] tests/test-umockdev.c: add slow_testbed_factor variable
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/umockdev/+bug/1831467
Origin: upstream, https://github.com/martinpitt/umockdev/pull/103

Similar to the previous patch to test-umockdev-record.vala,
this adjusts the test's acceptable -/+ range based on a factor.
---
 tests/test-umockdev.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

--- a/tests/test-umockdev.c
+++ b/tests/test-umockdev.c
@@ -42,6 +42,8 @@
 
 static gboolean has_real_sysfs;
 
+static int slow_testbed_factor = 1;
+
 typedef struct {
     UMockdevTestbed *testbed;
 } UMockdevTestbedFixture;
@@ -1738,14 +1740,14 @@ r 0 OK\n";
   close(fd);
 }
 
-/* assert that difference between first and second is "ms" milliseconds, ±20 ms */
+/* assert that difference between first and second is "ms" milliseconds, ±50 ms */
 static void
 assert_delta_t(const struct timeval * first, const struct timeval * second, int ms)
 {
     int delta = (second->tv_sec - first->tv_sec) * 1000;
     delta += (second->tv_usec - first->tv_usec) / 1000;
-    g_assert_cmpint(delta, >=, ms - 20);
-    g_assert_cmpint(delta, <=, ms + 20);
+    g_assert_cmpint(delta, >=, ms - 50 * slow_testbed_factor);
+    g_assert_cmpint(delta, <=, ms + 50 * slow_testbed_factor);
 }
 
 static void
@@ -2100,6 +2102,11 @@ t_testbed_proc(UMockdevTestbedFixture *
 int
 main(int argc, char **argv)
 {
+    const gchar *f = g_getenv("SLOW_TESTBED_FACTOR");
+
+    if (f != NULL && atoi(f) > 0)
+            slow_testbed_factor = atoi(f);
+
 #if !defined(GLIB_VERSION_2_36)
     g_type_init();
 #endif
